File: glEpisodes/src/main/java/com/kanawish/glepisodes/module/ScopeBuilder.java
package com.kanawish.glepisodes.module;

import android.app.Activity;
import android.app.Application;

import com.kanawish.glepisodes.module.app.AppModule;

import toothpick.Scope;
import toothpick.Toothpick;
import toothpick.smoothie.module.SmoothieActivityModule;
import toothpick.smoothie.module.SmoothieApplicationModule;

/**
 * We define our dependency injection scopes here.
 *
 */
public class ScopeBuilder {

    public static Scope buildApplicationScope(Application application) {
        Scope appScope = Toothpick.openScope(application);
        appScope.installModules(
                // SmoothieApplicationModule provides all the application-level system resource injection
                new SmoothieApplicationModule(application),
                new AppModule()
        );
        return appScope;
    }

    public static Scope buildActivityScope(Activity activity) {
        Scope scope = Toothpick.openScopes(activity.getApplication(), activity);
        scope.installModules(
                new SmoothieActivityModule(activity)
        );
        return scope;
    }

}

File: glEpisodes/src/main/java/com/kanawish/glepisodes/di/ScopeBuilder.kt
package com.kanawish.glepisodes.di

import android.app.Activity
import android.app.Application
import com.kanawish.di.ActivitySingleton

import toothpick.Scope
import toothpick.Toothpick
import toothpick.smoothie.module.SmoothieActivityModule
import toothpick.smoothie.module.SmoothieApplicationModule

/**
 *
 */
fun openApplicationScope(app: Application): Scope = Toothpick.openScope(app).apply {
    installModules(
            SmoothieApplicationModule(app),
            AppModule(app))
}

/**
 *
 */
fun openActivityScope(activity: Activity): Scope = Toothpick.openScopes(activity.application, activity).apply {
    // https://github.com/stephanenicolas/toothpick/wiki/Scope-Annotations
    bindScopeAnnotation(ActivitySingleton::class.java)

    installModules(
            SmoothieActivityModule(activity),
            ActivityModule(activity))
}
